src/pint/models/binary_piecewise.py

```diff
@@ -0,0 +1,465 @@
+        """
+        group_indexes = []
+        for i in range(0, len(self.binary_instance.piecewise_parameter_information)):
+            group_indexes.append(
+                self.binary_instance.piecewise_parameter_information[i][0]
+            )
+        return group_indexes
+
+    def get_group_indexes_in_four_digit_format(self):
+        """Get all the piecewise parameter labels in 4 digit format
+        Returns
+        -------
+        np.array
+                (length: number of piecewise groups) List of piecewise parameter labels in 4 digit format e.g with pieces T0X_0000, T0X_0001, T0X_0003, returns [0000,0001,0003]
+        """
+        group_indexes = []
+        for i in range(0, len(self.binary_instance.piecewise_parameter_information)):
+            group_indexes.append(
+                f"{int(self.binary_instance.piecewise_parameter_information[i][0]):04d}"
+            )
+        return group_indexes
+
+    def get_T0Xs_associated_with_toas(self, toas):
+        """Get a of all the piecewise T0s associated with TOAs
+        Parameters
+        ----------
+        toas :
+                Barycentric TOAs
+        Returns
+        -------
+        np.array
+                (length: toas) List of piecewise T0X values being used for each TOA
+        """
+        if hasattr(self.binary_instance, "group_index_array"):
+            temporary_storage = self.binary_instance.group_index_array
+        self.binary_instance.group_index_array = self.which_group_is_toa_in(toas)
+        barycentric_toa = self._parent.get_barycentric_toas(toas)
+        T0X_per_toa = self.binary_instance.piecewise_parameter_from_information_array(
+            toas
+        )[0]
+        if temporary_storage is not None:
+            self.binary_instance.group_index_array = temporary_storage
+        return T0X_per_toa
+
+    def get_A1Xs_associated_with_toas(self, toas):
+        """Get a of all the piecewise A1s associated with TOAs
+        Parameters
+        ----------
+        toas :
+                Barycentric TOAs
+        Returns
+        -------
+        np.array
+                (length: toas) List of piecewise A1X values being used for each TOA
+        """
+        if hasattr(self.binary_instance, "group_index_array"):
+            temporary_storage = self.binary_instance.group_index_array
+        self.binary_instance.group_index_array = self.which_group_is_toa_in(toas)
+        barycentric_toa = self._parent.get_barycentric_toas(toas)
+        A1X_per_toa = self.binary_instance.piecewise_parameter_from_information_array(
+            toas
+        )[1]
+        if temporary_storage is not None:
+            self.binary_instance.group_index_array = temporary_storage
+        return A1X_per_toa
+
+    def does_toa_reference_piecewise_parameter(self, toas, param):
+        """Query whether a TOA/list of TOAs belong(s) to a specific group
+        Parameters
+        ----------
+        toas :
```

Comment:
try to be explicit with the types/class references.  So this would be of type `pint.toa.TOAs`.

Reply:
I left some of these unchanged in the interest of time, since some of these functions are being removed